### PR TITLE
Check for native architecture and set GTSAM_COMPILE_OPTIONS_PUBLIC accordingly

### PR DESCRIPTION
**Pull request by stefangachter**
Currently, the flag `GTSAM_BUILD_WITH_MARCH_NATIVE` is limited to Apple architecture only. The commit checks, if the compiler supports `-march=native` and adds the option to `GTSAM_COMPILE_OPTIONS_PUBLIC` if `GTSAM_BUILD_WITH_MARCH_NATIVE` is on, see https://github.com/borglab/gtsam/pull/1138#pullrequestreview-1107033447

Note, could not test with Apple architecture, because hardware missing.

### CODE REVIEW

File: cmake/GtsamBuildTypes.cmake [hunk 1]
```diff
@@ -183,18 +183,33 @@ set(CMAKE_EXE_LINKER_FLAGS_PROFILING ${GTSAM_CMAKE_EXE_LINKER_FLAGS_PROFILING})
...
 endif()
 
 if (NOT MSVC)
-  option(GTSAM_BUILD_WITH_MARCH_NATIVE  "Enable/Disable building with all instructions supported by native architecture (binary may not be portable!)" OFF)
-  if(GTSAM_BUILD_WITH_MARCH_NATIVE AND (APPLE AND NOT CMAKE_SYSTEM_PROCESSOR STREQUAL "arm64"))
-    # Add as public flag so all dependant projects also use it, as required
-    # by Eigen to avid crashes due to SIMD vectorization:
-    list_append_cache(GTSAM_COMPILE_OPTIONS_PUBLIC "-march=native")
+  # Add as public flag so all dependant projects also use it, as required
+  # by Eigen to avid crashes due to SIMD vectorization:  option(GTSAM_BUILD_WITH_MARCH_NATIVE  "Enable/Disable building with all instructions supported by native architecture (binary may not be portable!)" OFF)
```
Comment (varunagrawal) [inline cmake/GtsamBuildTypes.cmake:194]: @stefangachter did you accidentally comment this flag out?

File: cmake/GtsamBuildTypes.cmake [hunk 2]
```diff
@@ -183,18 +183,33 @@ set(CMAKE_EXE_LINKER_FLAGS_PROFILING ${GTSAM_CMAKE_EXE_LINKER_FLAGS_PROFILING})
...
+  endif()
 endif()
 
 if (NOT MSVC)
-  option(GTSAM_BUILD_WITH_MARCH_NATIVE  "Enable/Disable building with all instructions supported by native architecture (binary may not be portable!)" OFF)
-  if(GTSAM_BUILD_WITH_MARCH_NATIVE AND (APPLE AND NOT CMAKE_SYSTEM_PROCESSOR STREQUAL "arm64"))
-    # Add as public flag so all dependant projects also use it, as required
-    # by Eigen to avid crashes due to SIMD vectorization:
-    list_append_cache(GTSAM_COMPILE_OPTIONS_PUBLIC "-march=native")
+  # Add as public flag so all dependant projects also use it, as required
```
Comment (varunagrawal) [inline cmake/GtsamBuildTypes.cmake:193]: Can you please undo this reorder of the comment? Or was there some reason to this that I am not seeing?

File: cmake/GtsamBuildTypes.cmake [hunk 3]
```diff
@@ -183,18 +183,33 @@ set(CMAKE_EXE_LINKER_FLAGS_PROFILING ${GTSAM_CMAKE_EXE_LINKER_FLAGS_PROFILING})
...
+  # by Eigen to avid crashes due to SIMD vectorization:  option(GTSAM_BUILD_WITH_MARCH_NATIVE  "Enable/Disable building with all instructions supported by native architecture (binary may not be portable!)" OFF)
+  if(APPLE AND (${CMAKE_CXX_COMPILER_ID} STREQUAL "Clang") AND ("${CMAKE_CXX_COMPILER_VERSION}" VERSION_LESS "15.0"))
+    if(GTSAM_BUILD_WITH_MARCH_NATIVE)
```
Comment (varunagrawal) [inline cmake/GtsamBuildTypes.cmake:196]: Can you please move this check to the outside since there is no `else` condition?

Something like the below is equivalent I believe

```cmake
if(GTSAM_BUILD_MARCH_NATIVE)
    if(APPLE AND (${CMAKE_CXX_COMPILER_ID} STREQUAL "Clang") AND ("${CMAKE_CXX_COMPILER_VERSION}" VERSION_LESS "15.0"))

        if(NOT CMAKE_SYSTEM_PROCESSOR STREQUAL "arm64")
            list_append_cache(GTSAM_COMPILE_OPTIONS_PUBLIC "-march=native")
        else()
            message(WARNING "The option GTSAM_BUILD_WITH_MARCH_NATIVE is ignored, because native architecture is not supported.")
        endif() # CMAKE_SYSTEM_PROCESSOR

    else()
        include(CheckCXXCompilerFlag)
        CHECK_CXX_COMPILER_FLAG("-march=native" COMPILER_SUPPORTS_MARCH_NATIVE)
        if(COMPILER_SUPPORTS_MARCH_NATIVE)
            list_append_cache(GTSAM_COMPILE_OPTIONS_PUBLIC "-march=native")
        else()
            message(WARNING "The option GTSAM_BUILD_WITH_MARCH_NATIVE is ignored, because native architecture is not supported.")
        endif()  # COMPILER_SUPPORTS_MARCH_NATIVE

    endif()
endif()  # GTSAM_BUILD_WITH_MARCH_NATIVE
```

Plus add some formatting, since this is a bit hard to parse?
Or preferably, use a [generator expression](https://cmake.org/cmake/help/latest/manual/cmake-generator-expressions.7.html)?

**Review by varunagrawal — CHANGES_REQUESTED** [review]
Some comments.

File: cmake/GtsamBuildTypes.cmake (hunk 3 above)
Comment (varunagrawal) [reply cmake/GtsamBuildTypes.cmake:196]: PS please correct me if my rearrangement of the if-blocks is wrong. I prefer generator expressions for this reason. 😅

File: cmake/GtsamBuildTypes.cmake (hunk 1 above)
Comment (stefangachter) [reply cmake/GtsamBuildTypes.cmake]: Sorry, my commit was pretty messy. -- Yes, commented out by mistakte. Corrected it.

File: cmake/GtsamBuildTypes.cmake (hunk 3 above)
Comment (stefangachter) [reply cmake/GtsamBuildTypes.cmake]: Indeed, Varun, your order makes much more sense. Sorry, I am not a cmake expert and do not master generator expressions. Thus, rewrote it with if-blocks. --- Again, I could not test it on Apple platform because of lack of hardware. Tested only with Ubuntu on Intel machine.

**Comment by varunagrawal** [comment]
@stefangachter I'll run this on my M1 mac tonight and if it works well, I'll approve. Thank you so much for your patience!

File: cmake/GtsamBuildTypes.cmake [hunk 4]
```diff
@@ -183,19 +183,35 @@ set(CMAKE_EXE_LINKER_FLAGS_PROFILING ${GTSAM_CMAKE_EXE_LINKER_FLAGS_PROFILING})
...
 
 if (NOT MSVC)
   option(GTSAM_BUILD_WITH_MARCH_NATIVE  "Enable/Disable building with all instructions supported by native architecture (binary may not be portable!)" OFF)
-  if(GTSAM_BUILD_WITH_MARCH_NATIVE AND (APPLE AND NOT CMAKE_SYSTEM_PROCESSOR STREQUAL "arm64"))
-    # Add as public flag so all dependant projects also use it, as required
-    # by Eigen to avid crashes due to SIMD vectorization:
-    list_append_cache(GTSAM_COMPILE_OPTIONS_PUBLIC "-march=native")
-  endif()
+  if(GTSAM_BUILD_WITH_MARCH_NATIVE)
+    if(APPLE AND (${CMAKE_CXX_COMPILER_ID} STREQUAL "Clang") AND ("${CMAKE_CXX_COMPILER_VERSION}" VERSION_LESS "15.0"))
```
Comment (varunagrawal) [inline cmake/GtsamBuildTypes.cmake:195]: Hey @stefangachter. Can you please explain why you're checking the version of the C++ compiler? I'm a bit lost on the reason here and it would be good to have the reason on record.

Comment (varunagrawal) [reply cmake/GtsamBuildTypes.cmake:195]: Found the [reason](https://stackoverflow.com/a/72597669/1236990).
AppleClang 15+ supports `march=native`.

**Comment by varunagrawal** [comment]
@stefangachter can you please check the box which says "Allow edits from maintainers"? It should be right below the unsubscribe button in the right column.

**Comment by stefangachter** [comment]
> @stefangachter can you please check the box which says "Allow edits from maintainers"? It should be right below the unsubscribe button in the right column.

@varunagrawal, it was already checked, if I am not mistaken:
image

**Comment by varunagrawal** [comment]
@stefangachter thanks! I was doing something silly.

**Review by varunagrawal — APPROVED** [review]
I made some minor updates using my CMake foo but kept the logic pretty much the same. Thanks for the major contribution @stefangachter!

**Comment by stefangachter** [comment]
Thank you, Varun! Wouldn't consider this a major contribution, but a contribution nevertheless :-)
